src/vs/workbench/services/runtimeSession/common/runtimeSession.ts

```diff
@@ -648,36 +652,51 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
 			return;
 		}
 
+		// Create a promise that resolves when the runtime is ready to use.
 		const startPromise = new DeferredPromise<string>();
+
+		// It's possible that startPromise is never awaited, so we log any errors here
+		// at the debug level since we still expect the error to be handled/logged elsewhere.
+		startPromise.p.catch(err => this._logService.debug(`Error starting runtime session: ${err}`));
+
+		// For notebook sessions, update the starting sessions map so that any concurrent
+		// start/restore requests return the same pending promise.
 		if (sessionMetadata.sessionMode === LanguageRuntimeSessionMode.Notebook) {
-			// Create a promise that resolves when the runtime is ready to use.
 			this._startingSessionsBySessionMapKey.set(sessionMapKey, startPromise);
 
-			// It's possible that startPromise is never awaited, so we log any errors here
-			// at the debug level since we still expect the error to be handled/logged elsewhere.
-			startPromise.p.catch((err) => this._logService.debug(`Error starting session: ${err}`));
-
 			this.setStartingSessionMaps(
 				sessionMetadata.sessionMode, runtimeMetadata, sessionMetadata.notebookUri);
 		}
+
+		try {
```

Comment:
This refactor was to fix two issues:

1. Make sure that `startPromise` is always returned instead of `throw`ing directly, so that it is awaited and we don't end up with uncaught errors.
2. Make sure that we always clear the starting session map when the promise resolves/rejects. There was a bug where if `doStartRuntimeSession` errored, the starting session map may not have been cleared.